crates/environ/src/fact/trampoline.rs

```diff
@@ -257,7 +380,294 @@ struct Memory<'a> {
+            format!("[adapter-callee]{}", adapter.name),
+        ));
+        self.instruction(I32Const(
+            i32::try_from(adapter.lower.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(RefFunc(adapter.callee.as_u32()));
+        self.instruction(I32Const(
+            i32::try_from(adapter.lift.instance.as_u32()).unwrap(),
+        ));
+        self.instruction(I32Const(param_count));
+        self.instruction(I32Const(result_count));
+        self.instruction(I32Const(EXIT_FLAG_ASYNC_CALLER));
+        self.instruction(Call(exit.as_u32()));
+
+        self.finish()
+    }
+
+    fn compile_async_start_adapter(
+        mut self,
+        adapter: &AdapterData,
+        sig: &Signature,
+        param_globals: Option<&[u32]>,
+    ) {
+        let mut temps = Vec::new();
+        let param_locals = if let Some(globals) = param_globals {
+            for global in globals {
+                let ty = self.module.globals[usize::try_from(*global).unwrap()];
+
+                self.flush_code();
+                self.module.funcs[self.result]
+                    .body
+                    .push(Body::GlobalGet(*global));
+                temps.push(self.local_set_new_tmp(ty));
+            }
+            temps
+                .iter()
+                .map(|t| (t.idx, t.ty))
+                .chain(if sig.results_indirect {
+                    sig.params
+                        .iter()
+                        .enumerate()
+                        .map(|(i, ty)| (i as u32, *ty))
+                        .last()
+                } else {
+                    None
+                })
+                .collect::<Vec<_>>()
+        } else {
+            sig.params
+                .iter()
+                .enumerate()
+                .map(|(i, ty)| (i as u32, *ty))
+                .collect::<Vec<_>>()
+        };
+
+        self.set_flag(adapter.lift.flags, FLAG_MAY_LEAVE, false);
+        self.translate_params(adapter, &param_locals);
+        self.set_flag(adapter.lift.flags, FLAG_MAY_LEAVE, true);
+
+        for tmp in temps {
+            self.free_temp_local(tmp);
+        }
+
+        self.finish();
+    }
+
+    fn compile_async_return_adapter(
+        mut self,
+        adapter: &AdapterData,
+        sig: &Signature,
+        result_globals: Option<&[u32]>,
+    ) {
+        let param_locals = sig
+            .params
+            .iter()
+            .enumerate()
+            .map(|(i, ty)| (i as u32, *ty))
+            .collect::<Vec<_>>();
+
+        self.set_flag(adapter.lower.flags, FLAG_MAY_LEAVE, false);
+        self.translate_results(adapter, &param_locals, &param_locals);
+        self.set_flag(adapter.lower.flags, FLAG_MAY_LEAVE, true);
+
+        if let Some(globals) = result_globals {
+            self.flush_code();
+            for global in globals {
+                self.module.funcs[self.result]
+                    .body
+                    .push(Body::GlobalSet(*global));
+            }
```

Comment:
Mind adding a helper method like `self.global_set(u32)` which automatically does the `flush_code` plus `Body::GlobalSet`?

Reply:
Are you concerned at all about the cost of calling `flush_code` for each `GlobalSet` rather than once for all of them?

Reply:
Oh, I see -- it doesn't matter because pushing the `GlobalSet`s won't add anything to `self.code`, so `flush_code` will be a no-op anyway.